pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/OneStageAuthenticationState.java

```diff
@@ -61,13 +80,47 @@ public AuthenticationDataSource getAuthDataSource() {
         return authenticationDataSource;
     }
 
+    /**
+     * Warning: this method is not intended to be called concurrently.
+     */
+    @Override
+    public CompletableFuture<AuthData> authenticateAsync(AuthData authData) {
+        if (authRole != null) {
+            // Authentication is already completed
+            return CompletableFuture.completedFuture(null);
+        }
+        this.authenticationDataSource = new AuthenticationDataCommand(
```

Comment:
I have one question:

When having `AuthChallenge`, do we need to renew `AuthenticationState`?

Reply:
Since this is a class for single stage authentication, there are no `AuthChallenges`. The primary reason to set this here instead of the constructor is that we set it with the `authData` that we authenticate. Note that this class always has `isExpired` return `false`, which means there are never any `AuthChallenges` generated.

If this were a multi-stage auth, this solution would not work.

Reply:
Your means that a single-stage authentication doesn't refresh authentication data, if so, when having `AuthChallenge`, we must new an `AuthenticationState`.

Reply:
> single-stage authentication doesn't refresh authentication data

Yes, that is what I mean. However, I don't agree that it means we need to create a new `AuthenticationState`.

My understanding of the the protocol is that the `AuthenticationState` method `isExpired` returns `true`, Pulsar will then call `refreshAuthentication()` to get the bytes to send to the client. The client responds with `AuthData`, and Pulsar passes the response to `authenticate(AuthData)`. Since `isExpired` is always false for `OneStateAuthenticationState`, there is no expectation of getting an `AuthChallenge` back. The single stage authentication is truly one stage.

Reply:
Sounds good, I see.